File: docs/src/rust/user-guide/expressions/folds.rs
use polars::prelude::*;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    
    // --8<-- [start:mansum]
    let df = df!(
        "a" => &[1, 2, 3],
        "b" => &[10, 20, 30],
    )?;

    let out = df
        .lazy()
        .select([fold_exprs(lit(0), |acc, x| Ok(Some(acc + x)), [col("*")]).alias("sum")])
        .collect()?;
    println!("{}", out);
    // --8<-- [end:mansum]

    // --8<-- [start:conditional]
    let df = df!(
        "a" => &[1, 2, 3],
        "b" => &[0, 1, 2],
    )?;

    let out = df
        .lazy()
        .filter(fold_exprs(
            lit(true),
            |acc, x| Some(acc.bitand(&x)),
            [col("*").gt(1)],
        ))
        .collect()?;
    println!("{}", out);
    // --8<-- [end:conditional]

    // --8<-- [start:string]
    let df = df!(
        "a" => &["a", "b", "c"],
        "b" => &[1, 2, 3],
    )?;

    let out = df
        .lazy()
        .select([concat_str([col("a"), col("b")], "")])
        .collect()?;
    println!("{:?}", out);
    // --8<-- [end:string]

    Ok(())
}

File: src/python/user-guide/expressions/folds.py
# --8<-- [start:setup]
import polars as pl

# --8<-- [end:setup]

# --8<-- [start:mansum]
df = pl.DataFrame(
    {
        "a": [1, 2, 3],
        "b": [10, 20, 30],
    }
)

out = df.select(
    pl.fold(acc=pl.lit(0), function=lambda acc, x: acc + x, exprs=pl.all()).alias(
        "sum"
    ),
)
print(out)
# --8<-- [end:mansum]

# --8<-- [start:conditional]
df = pl.DataFrame(
    {
        "a": [1, 2, 3],
        "b": [0, 1, 2],
    }
)

out = df.filter(
    pl.fold(
        acc=pl.lit(True),
        function=lambda acc, x: acc & x,
        exprs=pl.col("*") > 1,
    )
)
print(out)
# --8<-- [end:conditional]

# --8<-- [start:string]
df = pl.DataFrame(
    {
        "a": ["a", "b", "c"],
        "b": [1, 2, 3],
    }
)

out = df.select(pl.concat_str(["a", "b"]))
print(out)
# --8<-- [end:string]
